src/tools/rustfmt/src/parse/session.rs

```diff
@@ -231,7 +199,21 @@ impl ParseSess {
     }
 
     pub(crate) fn set_silent_emitter(&mut self) {
-        self.parse_sess.dcx = DiagCtxt::with_emitter(silent_emitter());
+        // Ideally this invocation wouldn't be necessary and the fallback bundle in
+        // `self.parse_sess.dcx` could be used.
+        let fallback_bundle = rustc_errors::fallback_fluent_bundle(
+            rustc_driver::DEFAULT_LOCALE_RESOURCES.to_vec(),
+            false,
+        );
+        // SAFETY: `fatal_dcx` will not be dropped while it is referenced by `self.parse_sess.dcx`,
+        // and `self.parse_sess.dcx` is written to by `ptr::write` so won't drop `fatal_dcx`.
+        unsafe {
+            let fatal_dcx = std::ptr::read(&self.parse_sess.dcx);
+            std::ptr::write(
+                &mut self.parse_sess.dcx,
+                DiagCtxt::with_emitter(Box::new(SilentEmitter { fallback_bundle, fatal_dcx, fatal_note: None }))
```

Comment:
You probably should wrap that `fatal_dcx` in a `ManuallyDrop`. In case `with_emitter` or the allocation panics, I believe this will result in the `dcx` being dropped twice.

Reply:
I might be misunderstanding, but wouldn't I just be immediately removing `fatal_dcx` from the `ManuallyDrop` to put it in the `SilentEmitter`, before the `with_emitter` call?

Reply:
I think that you are both correct. I.e., it looks to me like a `ManuallyDrop` around just the `fatal_dcx` local won't resolve things, but nonetheless the risk of a double-drop here is real...

Reply:
@davidtwco would you be willing to explore options here that don't require unsafe code? E.g. storing the `fatal_dcx` in an option and doing the none-swapping pattern, or something?

Reply:
Removed the unsafe now.